core/main/src/firebolt/handlers/provider_registrar.rs

```diff
@@ -103,138 +132,610 @@ impl ProviderRegistrar {
+            request,
+        );
+        Ok(ListenerResponse {
+            listening: listen,
+            event: context.method.clone(),
+        })
+    }
+
+    async fn callback_register_provider(
+        params: Params<'static>,
+        context: Arc<RpcModuleContext>,
+    ) -> Result<ListenerResponse, Error> {
+        info!("callback_register_provider: method={}", context.method);
+
+        if let Some(provides) = &context.provider_set.provides {
+            let mut params_sequence = params.sequence();
+            let call_context: CallContext = params_sequence.next().unwrap();
+            let request: ListenRequest = params_sequence.next().unwrap();
+            let listening = request.listen;
+
+            ProviderBroker::register_or_unregister_provider(
+                &context.platform_state,
+                provides.clone(),
+                context.method.clone(),
+                context.method.clone(),
+                call_context,
+                request,
+            )
+            .await;
+
+            Ok(ListenerResponse {
+                listening,
+                event: context.method.clone(),
+            })
+        } else {
+            Err(Error::Custom("Missing provides attribute".to_string()))
+        }
+    }
+
+    async fn callback_app_event_emitter(
+        params: Params<'static>,
+        context: Arc<RpcModuleContext>,
+    ) -> Result<Option<()>, Error> {
+        info!("callback_app_event_emitter: method={}", context.method);
+        if let Some(event) = &context.provider_set.provides_to {
+            let mut params_sequence = params.sequence();
+            let _call_context: CallContext = params_sequence.next().unwrap();
+            let result: Value = params_sequence.next().unwrap();
+
+            AppEvents::emit(&context.platform_state, event, &result).await;
+        }
+
+        Ok(None)
+    }
+
+    async fn callback_error(
+        params: Params<'static>,
+        context: Arc<RpcModuleContext>,
+    ) -> Result<Option<()>, Error> {
+        info!("callback_error: method={}", context.method);
+        let params_sequence = params.sequence();
+
+        if let Some(attributes) = context.provider_set.attributes {
+            if let Some(provider_response) = ProviderRegistrar::get_provider_response(
+                attributes.error_payload_type.clone(),
+                params_sequence,
+            ) {
+                ProviderBroker::provider_response(&context.platform_state, provider_response).await;
+            }
+        } else {
+            error!(
+                "callback_error: NO ATTRIBUTES: context.method={}",
+                context.method
+            );
+            return Err(Error::Custom(String::from("Missing provider attributes")));
+        }
+
+        Ok(None) as RpcResult<Option<()>>
+    }
+
+    async fn callback_provider_invoker(
+        params: Params<'static>,
+        context: Arc<RpcModuleContext>,
+    ) -> Result<Value, Error> {
```

Comment:
I'm not too sure about this. If we want to avoid mapping response data to specific types this needs to return a  `Value` so the receiver of the response can deserialize it as it knows the type. But does this clash with the way the firebolt schema is currently defined? Not sure how we want to handle this.

Reply:
Feel this could be a Value as long as we evaluate using the OpenRpc Validator for Request/Response and Event. Making Ripple an effective Pass through with Schema validation support,

Reply:
I think calls like `Content.requestUserInterest` expect the actual struct to be returned, e.g. in this case `EntityDetails` and not `Value::Object(EntityDetails)`, is that right @kpears201 ?

Reply:
...and if that's the case how will we ever get away from having to map response types like we still do in `ProviderRegistrar::get_provider_reponse()` ([here](https://github.com/rdkcentral/Ripple/blob/main/core/main/src/firebolt/handlers/provider_registrar.rs#L57))?

Reply:
Another point, for responses from APIs that have `x-provided-by`, ripple needs to be able to set the app ID parameter to the value of the app ID parameter (if it exists) in the provider response ([RPPL-2310](https://ccp.sys.comcast.net/browse/RPPL-2310)), meaning ripple needs to be able to process the response and not just pass-through.